Support Extra data models and subforms in validation

ivarne opened the pull request:
Work on supporting multiple data models and subforms in backend validations with expressions.

## Related Issue(s)
- #600 
- replaces https://github.com/Altinn/app-lib-dotnet/pull/701
- replaces https://github.com/Altinn/app-lib-dotnet/pull/619

## Main changes
The main goal for this PR is to add 

### New single interface for validation `IValidator`
With changes in one data element affects the validity of another element, the current hierarchy of validators and only `IFormDataValidator` supporting incremental validation break down. Adaptors for the old interfaces has been provided so no changes to current apps are required, but they might want to update in order to support incremental validation and read instance data with the `IInstanceDataAccessor`.

### New interface `IInstanceDataAccessor`
When validation has dependencies across data models, we need a cached way to access the content of data elements that does not require us to download the same data element multiple times for the same request. The implementation is not available from DI, but always passed as an argument in order to eliminate risk of it being reused cross requests because of Singleton services or other unintentionally shared global state.

### New `DataElementId` struct to ensure type safety for a wrapped Guid
36 places in the updated code need to reference the Guid of a data element, and adding type safety is an important step to ensure correctness. An implicit cast is provided from `DataElement` to simplify usage

### New `ModelBinding` struct in expressions engine
To support the new syntax for referring to a non default data element in components and expressions
```json
{
    "dataModelBindings": {"field": "path.to.data", "dataType": "default"}
    "hidden": ["equals", null, ["dataModel", "path.to.other", "prefill"]]
}
```

When handling references to a specific fields (eg. for removal) we needed to also add `DataReference` to combine a Field (path) with a specific DataElementId


### New PATCH api endpoint that allows frontend to update multiple data models at the same time
Saving and triggering incremental validation is expensive and some apps might want to batch multiple changes to reduce cost. The code in frontend would also be simplified if it could do a single request, instead of multiple.

### `ValidationIssueWithSource` class for external apis
Previously we had a single class `ValidationIssue` that included a `Source` property. Service owners sometimes tries to set this in their validators (because it shows up in intellisense), but as we override it based on `IValidator.ValidationSource` in our `ValidationService` it would be good if it wasn't there to confuse. This PR creates a new type `ValidationIssueWithSource` for external apis (where the source is included) and marks it `[Obsolete]` in the type returned in the validator interfaces.

### Respect new app logic settings for data elements `DisallowUserCreate` and `DisallowUserDelete`
This means that users can not create or delete this types. Typically used for prefill, when you want only the org token to be able to modify the number of data elements on an instance.

## Minor changes
Used IL code to generate real C# classes based on the json data models in shared tests. Previously different code was used to access properties of the test code as in production, which is not really good for tests. 

## Verification
- [ ] **Your** code builds clean without any errors or warnings
- [ ] Manual testing done (required)
- [ ] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [ ] All tests run green

## Documentation
- [ ] User documentation is updated with a separate linked PR in [altinn-studio-docs.](https://github.com/Altinn/altinn-studio-docs) (if applicable)

danielskovli committed:
Removes form-data delete check

danielskovli committed:
Implements `AllowUserCreate` and `AllowUserDelete` methodology. Fixes various tests and minor issues

danielskovli committed:
Temp project ref to Storage.Interface

danielskovli committed:
Access testing for DataController -> `Create` and `Delete`

danielskovli committed:
Verifies MaxCount before creating a new data element

danielskovli committed:
Merge branch 'main' into feature/subform

ivarne committed:
Implement support for multiple data models in expressions

ivarne committed:
Restructure validation to support incremental validation for multiple data models

ivarne committed:
Fix tests and cleanup

ivarne committed:
Fix sonar cloud issues

ivarne committed:
Add tests for shadow fields and hidden component data removal in process/next

HauklandJ committed:
chore: update to new storage nuget

ivarne committed:
Improve testing for multiple patch endpoint

ivarne committed:
Merge remote-tracking branch 'origin/main' into feat/multi-data-model-with-validation

ivarne committed:
Merge branch 'main' into feat/multi-data-model-with-validation

HauklandJ committed:
update Storage.Interface package reference to 3.33.0

danielskovli committed:
Merge remote-tracking branch 'origin/main' into feature/subform

martinothamar commented on src/Altinn.App.Core/Extensions/ServiceCollectionExtensions.cs:
```diff
@@ -322,7 +321,7 @@ private static void AddExternalApis(IServiceCollection services)
     private static void AddProcessServices(IServiceCollection services)
     {
-        services.TryAddTransient<IProcessEngine, ProcessEngine>();
+        services.TryAddScoped<IProcessEngine, ProcessEngine>();
```
Seems like this will break 1 app - digdir/oed

[DeceasedStateMiddleware](https://altinn.studio/repos/digdir/oed/src/commit/67bd185a198cbf380d72c173ae4fd73edc8ea953/App/Middlewares/DeceasedStateMiddleware.cs#L25) -> [StorageService](https://altinn.studio/repos/digdir/oed/src/commit/67bd185a198cbf380d72c173ae4fd73edc8ea953/App/Services/StorageService.cs#L38)

Since that type of middleware is a singleton, and they don't inject from the HttpContext. Should we just remember to notify them when we release? I don't think any of the other DI changes here affect anyone

ivarne replied:
I reverted the `AddScoped` changes. They are no longer needed after I rejected the idea of caching data element downloads using a scoped service.

I opened https://github.com/Altinn/app-lib-dotnet/issues/751, because I think we should default more to using scoped services.

danielskovli committed:
Fixes some tests that were double-creating the `default` data element

danielskovli committed:
Reorders DataController->MaxCount logic, adds missing ClassRef check before CreateAppModelData call

ivarne committed:
More changes to make tests work after merge

Make expression engine async

ivarne committed:
Merge remote-tracking branch 'origin/main' into feat/multi-data-model-with-validation

danielskovli committed:
Introduces `LayoutSet.Type` and adds null checks for `LayoutSet.Tasks`

ivarne committed:
Almost make the tests pass

ivarne committed:
Fix last remaining known bugs

ivarne committed:
Merge remote-tracking branch 'origin/main' into feat/multi-data-model-with-validation

ivarne committed:
fix async warning

ivarne committed:
Don't filter fields to remove by doing a double lookup

ivarne committed:
Merge remote-tracking branch 'origin/feature/subform' into feat/multi-data-model-with-validation

ivarne committed:
Fix sonar suggestions and tests

ivarne committed:
fix another warning

ivarne committed:
Fix another batch of sonar cloud issues

ivarne committed:
Implement SubFormComponent and use for validation

ivarne committed:
Merge remote-tracking branch 'origin/main' into feat/multi-data-model-with-validation

ivarne committed:
Fix tests after merge with main

ivarne committed:
Update storage.interfaces

ivarne committed:
Fix tests after storage upgrade

ivarne committed:
Add LayoutId to required validation descriptions

ivarne committed:
Add tests for subform validation

ivarne committed:
Add appMetadata to LayoutEvaluatorState

ivarne committed:
Fix typos and make sonar more happy

ivarne committed:
Revert changes to use scoped services

Opened #751 to consider making bigger changes for v9

ivarne committed:
Remove unused usings

ivarne committed:
Remove implicit cast, as it was no longer used in tests (and hid bugs in main code)

ivarne committed:
Remove file added by error

ivarne committed:
Merge remote-tracking branch 'origin/main' into feat/multi-data-model-with-validation

ivarne committed:
Merge remote-tracking branch 'origin/main' into feat/multi-data-model-with-validation

martinothamar commented on src/Altinn.App.Core/Internal/Validation/IValidationService.cs:
```diff
@@ -20,52 +20,31 @@ public interface IValidationService
 
     /// <summary>
-    /// Validates a single data element. Used by frontend to continuously validate form data as it changes.
+    ///
```
Would be nice with a doc here

ivarne replied:
Fixed.

martinothamar commented on src/Altinn.App.Core/Internal/Validation/ValidationService.cs:
```diff
@@ -26,251 +19,136 @@ public class ValidationService : IValidationService
-                )
-            )
-        );
+        var validators = _validatorFactory.GetValidators(taskId);
```
The comment here says "run", but the select isn't materialized

ivarne replied:
OK, comment reworded and moved

martinothamar commented on src/Altinn.App.Core/Models/Validation/ValidationIssueWithSource.cs:
```diff
@@ -0,0 +1,90 @@
+    /// <summary>
+    /// Converter function to create a <see cref="ValidationIssueWithSource"/> from a <see cref="ValidationIssue"/> and adding a source.
+    /// </summary>
+    public static ValidationIssueWithSource FromIssue(ValidationIssue issue, string source)
```
Really like this design! 🥂 
Could this (or even the whole class) be made internal?

ivarne replied:
It is part of the public api in controllers, so I don't think so.

martinothamar commented on src/Altinn.App.Api/Controllers/DataController.cs:
```diff
@@ -460,48 +519,76 @@ public async Task<ActionResult<DataPatchResponse>> PatchFormData(
             {
                 return Conflict(
-                    $"Cannot update data element of archived or deleted instance {instanceOwnerPartyId}/{instanceGuid}"
+                    new ProblemDetails()
```
I like that we use problem details more often, with the OTel setup this makes error responses a lot easier to debug 👏

martinothamar commented on src/Altinn.App.Api/Controllers/DataController.cs:
```diff
@@ -164,78 +162,89 @@ [FromQuery] string dataType
-                return await CreateAppModelData(org, app, instance, dataType);
+                return Conflict(
+                    $"Element type `{dataType}` has reached its maximum allowed count ({dataTypeFromMetadata.MaxCount})"
+                );
```
@danielskovli was this the error condition that was missing before? Can we make the response type `ProblemDetails` in that case?

danielskovli replied:
Yeah this is the check we discussed a while back 👍 

I think we can make the response whatever you/we want. I only did a 409 because it sort-of-almost followed the established pattern from this controller.

The frontend is specifically catching this response to display something other than "Oooops", but that check can easily be rewritten. It's not perfect as-is anyway, because it it assumes all 409 responses on `Create` is because of maxcount: https://github.com/Altinn/app-frontend-react/blob/feature/subform/src/features/subformData/useSubformMutations.tsx#L33

martinothamar replied:
Yeah I think 409 is good, I was only thinking about the string value being converted to `ProblemDetails` as we have middleware that annotates the ASP.NET Core trace with additional information when the response type is `ProblemDetails` (which means easier debugging)

danielskovli replied:
@martinothamar Unsurprisingly I don't know exactly what this should look like. But I can confirm that the frontend does check the body payload at all, only the response status code.

So if you're feeling inspired, please make whatever changes you think are necessary :)

martinothamar replied:
Pushed a commit for this

ivarne replied:
Thanks!

martinothamar commented on src/Altinn.App.Api/Controllers/DataController.cs:
```diff
@@ -452,6 +464,53 @@ public async Task<ActionResult<DataPatchResponse>> PatchFormData(
+        };
+        var response = await PatchFormDataMultiple(org, app, instanceOwnerPartyId, instanceGuid, request, language);
+
+        if (response.Result is OkObjectResult { Value: DataPatchResponseMultiple newResponse })
```
Is it possible to just check `response.Value` instead? The return from `PatchFormDataMultiple` might change to `StatusCode(200, ...)` for example which is also a valid `ActionResult<>` but then `.Result` is not this object. Feels kind of brittle

ivarne replied:
I'm open to suggestions, but I'm pretty sure my first instinct was to write `response.Value` and it did not work. Fortunately this code has tests, so if it breaks, we'll know about it. The correct way would probably be to add a helper method for both controller methods without the `ActionResult<>` mess.

martinothamar commented on src/Altinn.App.Api/Controllers/DataController.cs:
```diff
@@ -452,6 +464,53 @@ public async Task<ActionResult<DataPatchResponse>> PatchFormData(
+    [HttpPatch("")]
+    [ProducesResponseType(typeof(DataPatchResponseMultiple), 200)]
+    [ProducesResponseType(typeof(ProblemDetails), 409)]
+    [ProducesResponseType(typeof(ProblemDetails), 422)]
```
Codepath below produces 404 and 400 as well I think

ivarne replied:
Fixed

martinothamar commented on src/Altinn.App.Api/Controllers/DataController.cs:
```diff
@@ -556,23 +643,27 @@ [FromRoute] Guid dataGuid
+                dataType.AppLogic?.ClassRef is not null
+                && dataType.AppLogic.DisallowUserDelete
+                && !UserHasValidOrgClaim()
+            )
```
Is this a change in behavior, or does `DisallowUserDelete` default to `true`?

danielskovli replied:
Defaults to no value, eg `false`: https://github.com/Altinn/altinn-storage/blob/663f4b71747ab721b663d802c8506345be0eaa1f/src/Storage.Interface/Models/ApplicationLogic.cs#L48

martinothamar replied:
So in 
v8.3.5 - deleting form data is disallowed
v8.4.0 - deleting form data is allowed by default (so users who want same behavior as in v8.3.5 need to "opt in"?)

Has the change in default behavior been discussed? Not sure if this will affect anyone

danielskovli replied:
@martinothamar You could argue that `8.3.5` breaks with previous versions, and `8.4.0` restores the established behavior. 

The change has been discussed in the context of a subform, and agreed that this new implementation is desired.

Do as you will with that information 🙂

danielskovli replied:
Hmm I'm actually not sure about the versioning here, so don't read too much into that. But the essence of the situation is that I introduced a somewhat-breaking change (for API users), which Ivar has reverted by inverting the logic in DataController

martinothamar commented on src/Altinn.App.Api/Controllers/DataController.cs:
```diff
@@ -963,32 +1054,18 @@ await _instanceClient.UpdateDataValues(
-        }
-
-        return true;
+        return i?.Status?.Archived is null && i?.Status?.SoftDeleted is null && i?.Status?.HardDeleted is null;
```
With the `?` this reads a little funny. If `i` or `i.Status` are null then the instance is active?

martinothamar commented on src/Altinn.App.Core/Features/Telemetry.InterfaceFactory.cs:
```diff
@@ -0,0 +1,11 @@
+
+namespace Altinn.App.Core.Features;
+
+public partial class Telemetry
```
Daniel recently created the `Telemetry` folder

ivarne replied:
Oops, this file was from a different PR

ivarne committed:
Add ignoredValidators as a comma separated GET parameter to /validate endpoint

martinothamar commented on src/Altinn.App.Api/Controllers/ValidateController.cs:
```diff
@@ -67,9 +78,13 @@ public async Task<IActionResult> ValidateInstance(
         {
-            List<ValidationIssue> messages = await _validationService.ValidateInstanceAtTask(
+            var dataAccessor = new CachedInstanceDataAccessor(instance, _dataClient, _appMetadata, _appModel);
+            var ignoredSources = ignoredValidators?.Split(',').ToList();
```
Should we validate somewhere that these exist/are wellformed?

ivarne replied:
I don't think so. The requirement for these is that frontend can have a list of validators that it knows (thinks) it has a good client side implementation for. It would be lots of extra code to let frontend know if these validators are really used.

martinothamar commented on src/Altinn.App.Core/Features/Telemetry/Telemetry.Validation.cs:
```diff
@@ -18,52 +18,32 @@ private void InitValidation(InitContext context) { }
+        activity?.SetTag(
+            InternalLabels.ValidatorChangedElementsIds,
+            string.Join(',', changes.Select(c => $"{c.DataElement.DataType}:{c.DataElement.Id}"))
+        );
```
Should we use span events instead here? My current understanding is that tags/attributes are mostly suited for short/low-cardinality and thus filterable/indexable data while span events are must "structured logs"-like data colocated with the span

ivarne replied:
Done. I did not know it was possible.

martinothamar commented on src/Altinn.App.Core/Features/Validation/Default/ExpressionValidator.cs:
```diff
@@ -67,82 +100,109 @@ public async Task<List<ValidationIssue>> ValidateFormData(
+                if (
+                    hiddenFields.Exists(d =>
+                        d.DataElementId == dataElementId
+                        && resolvedField.Field.StartsWith(d.Field, StringComparison.InvariantCulture)
```
Is `StartsWith` correct here?

ivarne replied:
I'm pretty confident it is, because the only situation when a field would be a prefix of another field is if it is a repeating group that has a `hidden` or a `hiddenRow` property.

martinothamar commented on src/Altinn.App.Core/Internal/Expressions/LayoutEvaluatorStateInitializer.cs:
```diff
@@ -9,34 +12,101 @@ namespace Altinn.App.Core.Internal.Expressions;
     /// </summary>
-    public virtual Task<LayoutEvaluatorState> Init(
+    [Obsolete("Use the overload with ILayoutEvaluatorStateInitializer instead")]
+    public async Task<LayoutEvaluatorState> Init(
```
So if I understand correctly this and the code above is kept for compatibility, should we still have some tests for this to make sure it still works? From what I can tell it's only SSB that are using this to acccess `RemoveHiddenData`, so maybe just a test for that?

ivarne replied:
Yes, we should.

ivarne replied:
Added in aee32775a664c5117c2fd9f9bb5639f8fcea1de9

martinothamar commented on src/Altinn.App.Core/Models/Layout/PageComponentConverter.cs:
```diff
@@ -349,6 +374,19 @@ private BaseComponent ReadComponent(ref Utf8JsonReader reader, JsonSerializerOpt
+                //     showAddButton = reader.TokenType != JsonTokenType.False;
+                //     break;
+                // case "showdeletebutton":
+                //     showDeleteButton = reader.TokenType != JsonTokenType.False;
```
What are these commented lines for?

ivarne replied:
For the properties that frontend uses, but I realised I did not need for any purpose in backend. If we need them in the future it would be easier to uncomment than to write them from scratch.

ivarne committed:
Cleanup validatorService. Add new tests and telemetry

ivarne committed:
Merge remote-tracking branch 'origin/main' into feat/multi-data-model-with-validation

ivarne committed:
Make sonar cloud complaints editor warnings instead of suggestions

* Unused usings
* Mark members static

Much easier to see theese issues before sonar cloud runs

ivarne committed:
Activate CA1816: Call GC.SuppressFinalize correctly

Make it easier to get 0 sonar cloud warnings by making its issues warnings

ivarne committed:
Fix more sonar cloud issues

ivarne reviewed (commented):
Æsj, svar på kommentarer grupperte seg sammen til en review og ble ikke sendt

ivarne committed:
Make change to IProcessExclusiveGateway backwards compatible

ivarne committed:
Add additional status codes to swagger

martinothamar commented on src/Altinn.App.Core/Internal/Data/CachedFormDataAccessor.cs:
```diff
@@ -0,0 +1,159 @@
+///
+/// Do not add this to the DI container, as it should only be created explicitly because of data leak potential.
+/// </summary>
+internal sealed class CachedInstanceDataAccessor : IInstanceDataAccessor
```
Rename the file to match the class name?

ivarne replied:
😊

martinothamar commented on src/Altinn.App.Core/Internal/Data/CachedFormDataAccessor.cs:
```diff
@@ -0,0 +1,159 @@
+        public async Task<TValue> GetOrCreate(TKey key, Func<TKey, Task<TValue>> valueFactory)
+        {
+            Task<TValue> task;
+            lock (_cache)
```
What's the reasoning for using both a `ConcurrentDictionary` and a `lock`?

ivarne replied:
https://learn.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.addorupdate?view=net-8.0#system-collections-concurrent-concurrentdictionary-2-addorupdate(-0-system-func((-0-1))-system-func((-0-1-1)))
> If you call [AddOrUpdate](https://learn.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.addorupdate?view=net-8.0) simultaneously on different threads, addValueFactory may be called multiple times, but its key/value pair might not be added to the dictionary for every call.
For modifications and write operations to the dictionary, [ConcurrentDictionary<TKey,TValue>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2?view=net-8.0) uses fine-grained locking to ensure thread safety (read operations on the dictionary are performed in a lock-free manner). The addValueFactory and updateValueFactory delegates may be executed multiple times to verify the value was added or updated as expected. However, they are called outside the locks to avoid the problems that can arise from executing unknown code under a lock. Therefore, [AddOrUpdate](https://learn.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.addorupdate?view=net-8.0) is not atomic with regards to all other operations on the [ConcurrentDictionary<TKey,TValue>](https://learn.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2?view=net-8.0) class.

ivarne replied:
I suggested this in the maskinporten cache, but you wrote some wizard code that actually trigged that race condition. As I (think i) know the Lazy constructor is reasonably fast, there is no reason not to lock to be extra sure.

martinothamar replied:
Right, I just meant that you could use a normal `Dictionary<,>` since there is a "global" mutex wrapping the operations anyway

ivarne replied:
Yes I could, but ConcurrentDictionary sounds like exactly what I need.

Apart from needing Lazy<> and a lock.

ivarne replied:
Changed to using regular Dictionary, as it is protected by a lock

martinothamar commented on src/Altinn.App.Core/Internal/Process/ProcessTasks/Common/ProcessTaskFinalizer.cs:
```diff
@@ -54,58 +47,107 @@ public async Task Finalize(string taskId, Instance instance)
+                    Guid.Parse(dataElement.Id)
+                );
+            })
+        );
```
I think the other places `CachedInstanceDataAccessor` is used, the cache is updated after updates/modifications, but not here, maybe because it isn't needed. If it is true that when using `CachedInstanceDataAccessor` and making data updates, we should always make sure the cache is in sync, can we write the code such that updating data always updates  the cache or vice versa?

ivarne replied:
If you just mutate the data models, the cache will be in sync. The issue is if you add or remove data elements from the instance, or if you want to replace the root data object in the cache.

We actually have such a modification when removing shadow fields (or making a copy of the data element without shadow fields)

https://github.com/Altinn/app-lib-dotnet/blob/450f53d978f52e65c275b8146d4c6ac6ebea1c55/src/Altinn.App.Core/Internal/Process/ProcessTasks/Common/ProcessTaskFinalizer.cs#L157-L185

Maybe `IInstanceDataProcessor` should have apis to mutate Instance.Data and sync with both local state and storage?

ivarne replied:
This has been resolved by splitting the interface into `IInstanceDataAccessor` and `IInstanceDataMutator`. One for validations, and the mutator for actions and data processing.
 CachedInstanceDataAccessor implements both and can be used in both

ivarne committed:
Update Instance.Data when writing copied data without shadow fields

ivarne committed:
Merge remote-tracking branch 'origin/main' into feat/multi-data-model-with-validation

ivarne committed:
Add concept of nonIncrementalValidators

Some validators can neither run on every request for efficeincy,
nor has a sensible way to implement HasRelevantChanges.

Theese validators need to only run on process/next,
or when explicitly requested in a call to /validate

Also added NoIncrementalUpdates (with [JsonIgnore(WhenWritingDefault)]) to
ValidationIssueWithSource, so that FE can distinguish
what issues won't be fixed with incremental updates.

/validate now supports a boolean to run either only incremental, only non
incremnetal or all validators (default)

ivarne committed:
Add ModelSerializationService and other cleanup as a prerequisite to new dataProcessor

ivarne committed:
Add draft support for adding/removing/modifying other data elements in dataProcessWrite

ivarne committed:
Code review fixes

ivarne committed:
Various cleanup tasks

* Use IInstanceDataMutator for actions
* Return List instead of dictionary for models and validation issues in MultipleDataPatch for improved swagger doc
* New IDataWriteProcessor interface with IInstanceDataMutator and change list to mimik IValidator
* IValidator gets ShouldRunForTask(string taskId) so that
    * Required validators only runs for tasks with layout.
    * Expression validators only runs for tasks with datatypes that has expressions validation files
* Revert DataElementIdentifier to a simple wrapper for Guid (that caches the string version)

ivarne committed:
Rename DataElementId to DataElementIdentifier

ivarne committed:
Merge pull request #809 from Altinn/feat/ivarne/multiDataProcessing

Add IInstanceDataModifier and use it for a new dataProcessing interface and extend the UserActionContext to support modifying data elements.

ivarne committed:
Merge remote-tracking branch 'origin/main' into feat/multi-data-model-with-validation

ivarne committed:
Fix issues and tests failing after merge

ivarne committed:
Rename id=>dataElementId in api

also a few final touches and possible test fixes

ivarne committed:
Merge remote-tracking branch 'origin/main' into feat/multi-data-model-with-validation

ivarne committed:
Use async versions of File access methods in DataClientMock

ivarne committed:
Reorganize ProcessTaskFinalizer to use CachedInstanceDataAccessor

martinothamar committed:
Use 'ProblemDetails' for new 409 response branch in DataController

ivarne committed:
Various fixes and api cleanup

* Refactor DataController and separate out DataElementAccessChecker so that it is easier to reuse logic and return consistent ProblemDetails across APIs
* Fix error in compatibility code with ActionController not returning data from UpdatedDataModels when IInstanceDataMutator is not used.
* Return Instance from UserAction endpoint in case data elements were added or removed.
* Check that instance is active when updating metadata.
* Use PatchListItem in multiplePatchEndpoint instead of dictionary for better OpenApi documentation

ivarne committed:
Fix sonar issues

ivarne committed:
Use new DataProcessing interface in PUT

* use IInstanceMutator
* Make IInstanceMutator update dataValues and PresentationTexts and remove from places that uses it

ivarne committed:
Merge remote-tracking branch 'origin/main' into feat/multi-data-model-with-validation

ivarne committed:
Fix formatting and move the .gitignore file protecting test data

ivarne committed:
Fix codeQL warning

ivarne committed:
Add test for FormDataValidatorWrapper

ivarne committed:
Merge branch 'main' into feat/multi-data-model-with-validation

HauklandJ commented on src/Altinn.App.Core/Helpers/Serialization/ModelSerializationService.cs:
```diff
@@ -0,0 +1,265 @@
+            return serializer.Deserialize(xmlTextReader)
+                ?? throw new InvalidOperationException("Deserialization returned null");
+        }
+        catch (InvalidOperationException)
```
I stedet for å bruke exceptions til å styre control flow, kunne vi gjort noe som:
```
    bool containsNamespace = CheckIfNamespaceExists(streamContent);
    XmlSerializer serializer = containsNamespace
        ? _xmlSerializer.GetSerializer(modelType)
        : _xmlSerializer.GetSerializerIgnoreNamespace(modelType);
```

ivarne replied:
Jeg skulle gjerne gjort det, men jeg aner ikke hvordan jeg implementerer `CheckIfNamespaceExists(stream)`, og dette er kode jeg har kopiert fra slik vi gjorde ting tidligere.

HauklandJ commented on src/Altinn.App.Core/Helpers/Serialization/ModelSerializationService.cs:
```diff
@@ -0,0 +1,265 @@
+        string streamContent = Encoding.UTF8.GetString(data.RemoveBom());
+        if (string.IsNullOrWhiteSpace(streamContent))
+        {
+            throw new Exception("No XML content read from stream");
```
Bruke ArgumentException?

ivarne replied:
Vet ikke helt om jeg ser poenget gitt at vi er enige om å ikke styre controlFlow med exceptions.
```suggestion
            throw new ArgumentException("No XML content read from stream");
```

HauklandJ reviewed (approved):
LGTM, some minor things to consider

ivarne committed:
Update src/Altinn.App.Core/Helpers/Serialization/ModelSerializationService.cs

ivarne committed:
Add test for legacy LayoutEvaluatorInitializer
